File: hex_terrain_editor/editor/select_tool.py
"""Select tool for selecting hexes for batch operations."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..utils.hex_math import pixel_to_hex

if TYPE_CHECKING:
    from ..model.hex_grid import HexGrid


class SelectTool:
    """Click to select individual hexes, or drag to select multiple."""

    def __init__(self, grid: "HexGrid", on_selection_changed: Any = None) -> None:
        self.grid = grid
        self.on_selection_changed = on_selection_changed
        self.selected: set[tuple[int, int]] = set()
        self._dragging = False

    def activate(self) -> None:
        pass

    def deactivate(self) -> None:
        self.selected.clear()
        self._dragging = False

    def on_press(self, world_x: float, world_y: float) -> None:
        q, r = pixel_to_hex(world_x, world_y, self.grid.hex_size)
        cell = self.grid.get(q, r)
        if cell is not None:
            self.selected = {(q, r)}
            self._dragging = True
            if self.on_selection_changed:
                self.on_selection_changed(self.selected)

    def on_move(self, world_x: float, world_y: float) -> None:
        if self._dragging:
            q, r = pixel_to_hex(world_x, world_y, self.grid.hex_size)
            cell = self.grid.get(q, r)
            if cell is not None:
                self.selected.add((q, r))
                if self.on_selection_changed:
                    self.on_selection_changed(self.selected)

    def on_release(self, world_x: float, world_y: float) -> None:
        self._dragging = False

    def clear_selection(self) -> None:
        self.selected.clear()
        if self.on_selection_changed:
            self.on_selection_changed(self.selected)

File: src/editor/select_tool.rs
use std::collections::HashSet;
use crate::hex_math;
use crate::model::hex_grid::HexGrid;

pub struct SelectTool {
    pub selected: HashSet<(i32, i32)>,
    dragging: bool,
    start_hex: Option<(i32, i32)>,
}

impl SelectTool {
    pub fn new() -> Self {
        Self {
            selected: HashSet::new(),
            dragging: false,
            start_hex: None,
        }
    }

    pub fn on_press(&mut self, world_x: f32, world_y: f32, grid: &HexGrid, additive: bool) {
        let (q, r) = hex_math::pixel_to_hex(world_x, world_y, grid.hex_size);
        self.dragging = true;
        self.start_hex = Some((q, r));

        if !additive {
            self.selected.clear();
        }

        if grid.get(q, r).is_some() {
            self.selected.insert((q, r));
        }
    }

    pub fn on_move(&mut self, world_x: f32, world_y: f32, grid: &HexGrid) {
        if !self.dragging {
            return;
        }
        let (q, r) = hex_math::pixel_to_hex(world_x, world_y, grid.hex_size);
        if grid.get(q, r).is_some() {
            self.selected.insert((q, r));
        }
    }

    pub fn on_release(&mut self) {
        self.dragging = false;
        self.start_hex = None;
    }

    pub fn clear_selection(&mut self) {
        self.selected.clear();
    }

    pub fn is_selected(&self, q: i32, r: i32) -> bool {
        self.selected.contains(&(q, r))
    }
}

impl Default for SelectTool {
    fn default() -> Self {
        Self::new()
    }
}
